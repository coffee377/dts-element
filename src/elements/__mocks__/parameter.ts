export class Parameter {

  public parameters: any;

  constructor(parameters: any) {
    this.parameters = parameters;
  }

  public _emit(): string {
    return `[Parameter ${this.parameters.name}]`;
  }

}
