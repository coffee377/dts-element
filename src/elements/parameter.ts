import {ParameterFlag} from '../collections';
import {any_type} from '../constants';
import {AnyElement, Element} from '../element';
import {emit_parameter_main} from '../helpers/emit-parameter-main';
import {emit_parameter_type} from '../helpers/emit-parameter-type';
import {AnyType} from './type';

export interface IParameterRequiredParameters {
  name: string;
}

export interface IParameterOptionalParameters {
  type: AnyType;
  flag: ParameterFlag;
}

export class Parameter extends Element<IParameterRequiredParameters, IParameterOptionalParameters> {

  public get default_parameters(): IParameterOptionalParameters {
    return {
      type: any_type,
      flag: null,
    };
  }

  public _emit(container: AnyElement | null): string {
    const {name, type, flag} = this.parameters;
    return `${emit_parameter_main(name, flag)}: ${emit_parameter_type(type, flag, container)}`;
  }

}
