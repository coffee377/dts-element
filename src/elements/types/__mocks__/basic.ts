export class BasicType {

  public parameters: any;

  constructor(parameters: any) {
    this.parameters = parameters;
  }

  public _emit(): string {
    return `[BasicType ${this.parameters.name}]`;
  }

}
