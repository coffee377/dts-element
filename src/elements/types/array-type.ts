import {emit_brackets} from '../../helpers/emit-brackets';
import {Stack} from '../../stack';
import {Type} from '../type';

export interface IArrayTypeRequiredParameters {
  owned: Type;
}

// tslint:disable-next-line no-empty-interface
export interface IArrayTypeOptionalParameters {}

/**
 * ```ts
 * type X = some_type[];
 * //       ^^^^^^^^^^^
 * ```
 */
export class ArrayType extends Type<IArrayTypeRequiredParameters, IArrayTypeOptionalParameters> {

  private _instance_of_array_type: true;

  public get default_parameters(): IArrayTypeOptionalParameters {
    return {};
  }

  public _emit(stack: Stack): string {
    const {owned} = this.parameters;
    return `${emit_brackets(owned.emit(stack), owned)}[]`;
  }

}
