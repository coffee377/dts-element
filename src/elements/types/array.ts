import {AnyElement} from '../../element';
import {AnyType, Type} from '../type';

export interface IArrayTypeRequiredParameters {
  owned: AnyType;
}

// tslint:disable-next-line no-empty-interface
export interface IArrayTypeOptionalParameters {}

export class ArrayType extends Type<IArrayTypeRequiredParameters, IArrayTypeOptionalParameters> {

  public get default_parameters(): IArrayTypeOptionalParameters {
    return {};
  }

  public _emit(container: AnyElement | null): string {
    return `${this.parameters.owned._emit(container)}[]`;
  }

}
