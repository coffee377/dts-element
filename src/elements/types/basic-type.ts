import {Stack} from '../../stack';
import {Type} from '../type';

export interface IBasicTypeRequiredParameters {
  name: string;
}

// tslint:disable-next-line no-empty-interface
export interface IBasicTypeOptionalParameters {}

export class BasicType extends Type<IBasicTypeRequiredParameters, IBasicTypeOptionalParameters> {

  public get default_parameters(): IBasicTypeOptionalParameters {
    return {};
  }

  public _emit(_stack: Stack): string {
    return this.parameters.name;
  }

}
