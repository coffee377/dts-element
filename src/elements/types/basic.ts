import {AnyElement} from '../../element';
import {Type} from '../type';

export interface IBasicTypeRequiredParameters {
  name: string;
}

// tslint:disable-next-line no-empty-interface
export interface IBasicTypeOptionalParameters {}

export class BasicType extends Type<IBasicTypeRequiredParameters, IBasicTypeOptionalParameters> {

  public get default_parameters(): IBasicTypeOptionalParameters {
    return {};
  }

  public _emit(_container: AnyElement | null): string {
    return this.parameters.name;
  }

}
