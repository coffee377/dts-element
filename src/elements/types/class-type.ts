import {emit_declaration_type_name} from '../../helpers/emit-declaration-type-name';
import {emit_generics} from '../../helpers/emit-generics';
import {Stack} from '../../stack';
import {ClassDeclaration} from '../declarations/class-declaration';
import {Type} from '../type';

export interface IClassTypeRequiredParameters {
  owned: string | ClassDeclaration;
}

export interface IClassTypeOptionalParameters {
  generics: Type[];
}

export class ClassType extends Type<IClassTypeRequiredParameters, IClassTypeOptionalParameters> {

  public get default_parameters(): IClassTypeOptionalParameters {
    return {
      generics: [],
    };
  }

  public _emit(stack: Stack): string {
    const name = emit_declaration_type_name(this.parameters.owned);
    return `${name}${emit_generics(this.parameters.generics, stack)}`;
  }

}
