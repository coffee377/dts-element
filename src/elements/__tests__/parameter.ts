jest.unmock('../parameter');

import {Parameter} from '../parameter';
import {BasicType} from '../types/basic';

describe('#emit()', () => {
  it('should return correctly', () => {
    expect(new Parameter({
      name: 'param',
      type: new BasicType({name: 'Param'}),
    }).emit()).toMatchSnapshot();
  });
});
