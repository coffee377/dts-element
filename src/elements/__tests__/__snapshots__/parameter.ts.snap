// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#emit() should return correctly 1`] = `"[emit-parameter-main param]: [emit-parameter-type [BasicType Param]]"`;
